tensorboard/plugins/pr_curve/tf_pr_curve_dashboard/tf-pr-curve-card.ts

```diff
@@ -0,0 +1,160 @@
+/* tslint:disable:no-namespace variable-name */
+
+"use strict";
+import {Canceller} from "../tf-backend/canceller.js";
+import {getRouter} from "../tf-backend/router.js";
+import {runsColorScale} from "../tf-color-scale/colorScale.js";
+import * as ChartHelpers from '../vz-line-chart/vz-chart-helpers.js';
+
+interface PrCurvePoint {
+  scalar: number;
+  recall: number;
+}
+
+Polymer({
+  is: "tf-pr-curve-card",
+  properties: {
+    runs: Array,
+    tag: String,
+
+    // The card will display the PR curve at this step or the one closest to it,
+    // but less than it.
+    stepCap: Number,
+    
+    _runToData: Object,
+    /** @type {Function} */
+    _colorScaleFunction: {
+      type: Object,  // function: string => string
+      value: () => ({scale: runsColorScale}),
+    },
+    _runColor: {
+      type: String,
+      computed: "_computeRunColor(run)",
+    },
+    requestManager: Object,
+    _canceller: {
+      type: Object,
+      value: () => new Canceller(),
+    },
+    _attached: Boolean,
+    // The value field is a function that returns a function because Polymer
+    // will actually call the value field if the field is a function. However,
+    // we actually want the value itself to be a function.
+    _xComponentsCreationMethod: {
+      type: Object,
+      value: () => (() => {
+        const scale = new Plottable.Scales.Linear();
+        return {
+          scale: scale,
+          axis: new Plottable.Axes.Numeric(scale, 'bottom'),
+          accessor: (d: PrCurvePoint) => d.recall,
+        };
+      }),
+      readOnly: true,
+    }
+  },
+  observers: [
+      "reload(run, tag)",
+      "_runsChanged(_attached, runs.*)",
+      "_setChartData(_runToData, stepCap)",
+  ],
+  _computeRunColor(run) {
+    return this._colorScaleFunction(run);
+  },
+  _serializedRunToPrCurveData(runToPrCurveData) {
+    return JSON.stringify(runToPrCurveData);
+  },
+  attached() {
+    // Defer reloading until after we're attached, because that ensures that
+    // the requestManager has been set from above. (Polymer is tricky
+    // sometimes)
+    this._attached = true;
+    this.reload();
+  },
+  reload() {
+    if (!this._attached) {
+      return;
+    }
+    this._canceller.cancelAll();
+    const router = getRouter();
+    let url = router.pluginRoute("pr_curve", "/pr_curves");
+    url += url.indexOf('?') > -1 ? '&' : '?';
+    url += `tag=${this.tag}&runs=${this.runs.join(',')}`;
+    const updateData = this._canceller.cancellable(result => {
+      if (result.cancelled) {
+        return;
+      }
+
+      // This is a mapping between run and data.
+      const runToData = result.value;
+      this.set('_runToData', runToData);
+      this.fire('data-updated', {
+        'runToData': runToData,
+      });
+    });
+    this.requestManager.request(url).then(updateData);
+  },
+  _setChartData(runToData, stepCap) {
+    if (!runToData || !_.isNumber(stepCap)) {
```

Comment:
Polymer observers are not invoked before all of their arguments become non-`undefined`. In what circumstances would each of these checks trigger?